accounts/views.py

```diff
@@ -98,10 +100,177 @@ def signup(request):
         form = CustomUserCreationForm()
         return render(request, "accounts/login.html", {"form": form})
 
+
 def settings_view(request):
     return render(request, "accounts/profile_settings.html")
 
+
 def onboarding(request):
     # if(request.session.get("onboarding")):
     return render(request, "accounts/onboarding.html")
 
+
+# --- Friend handling begins ---
+
+
+def send_friend_request(request):
+    """
+    This function handles when a user wants to send a friend request to another user.
+    """
+    # check request is POST and has username field
+    if request.method != "POST":
+        return redirect("landing_page:index")
+    try:
+        data = json.loads(request.body)
+        request_username = data["username"]
+    except (json.JSONDecodeError, KeyError):
+        return HttpResponse("Invalid JSON", status=400)
+    print("request_username: ", request_username)
+    # check user is logged in and User exists
+    if not request.user.is_authenticated:
+        return redirect("accounts:login")
+    try:
+        receiver_user = User.objects.get(username=request_username)
+    except User.DoesNotExist:
+        return HttpResponse("User not found.", status=404)
+    receiver_id = receiver_user.pk
+    # check if user is trying to send a friend request to themselves
+    if request.user.pk == receiver_id:
+        return HttpResponse("You can't send a friend request to yourself.", status=400)
+    # check if user is already friends with the other user
+    if request.user.friends.filter(pk=receiver_id).exists():
+        return HttpResponse("You are already friends with this user.", status=400)
+    # check if user has already sent a friend request to the other user
+    if FriendRequest.objects.filter(
+        sender=request.user, receiver=receiver_user
+    ).exists():
+        return HttpResponse(
+            "You have already sent a friend request to this user.", status=400
+        )
+    # check if user has already received a friend request from the other user
+    if FriendRequest.objects.filter(
+        sender=receiver_user, receiver=request.user
+    ).exists():
+        return HttpResponse(
+            "You have already received a friend request from this user.", status=400
```

Comment:
hmmm should all of these just be 'bad request'?
like, isn't a friend request already existing kind of a 409 - conflict issue instead?

the one thing i liked about ben's 337 was learning about https://http.cat/, i know you've taken 337 too so maybe you've seen it. not sure if all of these should be 400.

Reply:
Benny D 🐐🐐🐐